### PR TITLE
Add a blob-specific cache priority (#10461)

### PR DESCRIPTION
**Pull request by Connor1996**
Summary:
RocksDB's `Cache` abstraction currently supports two priority levels for items: high (used for frequently accessed/highly valuable SST metablocks like index/filter blocks) and low (used for SST data blocks). Blobs are typically lower-value targets for caching than data blocks, since 1) with BlobDB, data blocks containing blob references conceptually form an index structure which has to be consulted before we can read the blob value, and 2) cached blobs represent only a single key-value, while cached data blocks generally contain multiple KVs. Since we would like to make it possible to use the same backing cache for the block cache and the blob cache, it would make sense to add a new, lower-than-low cache priority level (bottom level) for blobs so data blocks are prioritized over them.

This task is a part of https://github.com/facebook/rocksdb/issues/10156

Pull Request resolved: https://github.com/facebook/rocksdb/pull/10461

Reviewed By: siying

Differential Revision: D38672823

Pulled By: ltamasi

fbshipit-source-id: 90cf7362036563d79891f47be2cc24b827482743

### CODE REVIEW

**Review by tonyxuqqi — APPROVED** [review]

File: table/block_based/block_based_table_reader_test.cc [hunk 1]
```diff
@@ -163,7 +163,7 @@ TEST_P(BlockBasedTableReaderTest, MultiGet) {
         // Internal key is constructed directly from this key,
         // and internal key size is required to be >= 8 bytes,
         // so use %08u as the format string.
-        sprintf(k, "%08u", key);
+        snprintf(k, sizeof(k), "%08u", key);
```
Comment (v01dstar) [inline table/block_based/block_based_table_reader_test.cc:166]: Were these changes for debugging only? If so, maybe we should consider reverting, unless we want to contribute this PR to the facebook upstream

**Review by v01dstar — APPROVED** [review]
LGTM

File: cache/lru_cache.cc [hunk 2]
```diff
@@ -263,10 +297,24 @@ void LRUCacheShard::MaintainPoolSize() {
...
     lru_low_pri_->SetInHighPriPool(false);
+    lru_low_pri_->SetInLowPriPool(true);
     size_t total_charge =
         lru_low_pri_->CalcTotalCharge(metadata_charge_policy_);
     assert(high_pri_pool_usage_ >= total_charge);
     high_pri_pool_usage_ -= total_charge;
+    low_pri_pool_usage_ += total_charge;
+  }
+
+  while (low_pri_pool_usage_ > low_pri_pool_capacity_) {
```
Comment (tonyxuqqi) [inline cache/lru_cache.cc:308]: I don't understand that why the actual link is not updated, but just update the usage_?
The entry is still with the previous priority link (either lru_ or lru_low_pri_

File: cache/lru_cache.cc [hunk 3]
```diff
@@ -263,10 +297,24 @@ void LRUCacheShard::MaintainPoolSize() {
...
+    // Overflow last entry in low-pri pool to bottom-pri pool.
+    lru_bottom_pri_ = lru_bottom_pri_->next;
+    assert(lru_bottom_pri_ != &lru_);
```
Comment (tonyxuqqi) [inline cache/lru_cache.cc:311]: should it be lru_low_pri_ for lru_?

File: table/block_based/block_based_table_reader_test.cc (hunk 1 above)
Comment (Connor1996) [reply table/block_based/block_based_table_reader_test.cc:166]: Newer compiler complains safety issue

File: cache/lru_cache.cc (hunk 3 above)
Comment (Connor1996) [reply cache/lru_cache.cc:311]: Nope, same as upstream https://github.com/facebook/rocksdb/pull/10461/files#diff-1b304e22abe7464a3675347f930d19e35a77e12f9e8d590a9e5c79dd7f61398bR309

File: cache/lru_cache.cc (hunk 2 above)
Comment (Connor1996) [reply cache/lru_cache.cc:308]: Three priorities are composed in one link, like below
```
A---->B---->C---->D---->E---->F
      ^           ^           ^
      |           |           |
   bottom        low         high
```
So when one priority is full, only need to move the pointer of lower priority

**Review by glorv — APPROVED** [review]

**Pull request merged**